mono/metadata/mono-perfcounters.c

```diff
@@ -1483,231 +1572,292 @@ typedef struct {
 {
+	HANDLE_FUNCTION_ENTER ()
+
 	MonoError error;
+	error_init (&error);
+
 	int result = FALSE;
-	int i, size;
-	int num_counters = mono_array_length (items);
-	int counters_data_size;
+	int const num_counters = mono_array_handle_length (items);
 	char *name = NULL;
 	char *chelp = NULL;
 	char **counter_info = NULL;
-	char *p;
-	SharedCategory *cat;
+	gboolean unlock = FALSE;
 
 	/* FIXME: ensure there isn't a category created already */
-	name = mono_string_to_utf8_checked (category, &error);
-	if (!mono_error_ok (&error))
-		goto failure;
-	chelp = mono_string_to_utf8_checked (help, &error);
-	if (!mono_error_ok (&error))
-		goto failure;
+	name = mono_string_handle_to_utf8 (category, &error);
+	goto_if_nok (&error, failure);
+
+	chelp = mono_string_handle_to_utf8 (help, &error);
+	goto_if_nok (&error, failure);
+
 	counter_info = g_new0 (char*, num_counters * 2);
 	/* calculate the size we need structure size + name/help + 2 0 string terminators */
+	int size;
 	size = G_STRUCT_OFFSET (SharedCategory, name) + strlen (name) + strlen (chelp) + 2;
+
+	int i;
 	for (i = 0; i < num_counters; ++i) {
-		CounterCreationData *data = mono_array_get (items, CounterCreationData*, i);
-		counter_info [i * 2] = mono_string_to_utf8_checked (data->name, &error);
-		if (!mono_error_ok (&error))
-			goto failure;
-		counter_info [i * 2 + 1] = mono_string_to_utf8_checked (data->help, &error);
-		if (!mono_error_ok (&error))
-			goto failure;
+		MONO_HANDLE_LOCAL_VARIABLE_INITIALIZED_NULL (CounterCreationData, data);
+		MONO_HANDLE_ARRAY_GETREF (data, items, i);
+		counter_info [i * 2] = mono_string_handle_to_utf8 (MONO_HANDLE_NEW_GET (MonoString, data, name), &error);
+		goto_if_nok (&error, failure);
+		counter_info [i * 2 + 1] = mono_string_handle_to_utf8 (MONO_HANDLE_NEW_GET (MonoString, data, help), &error);
+		goto_if_nok (&error, failure);
 		size += sizeof (SharedCounter) + 1; /* 1 is for the help 0 terminator */
 	}
+
 	for (i = 0; i < num_counters * 2; ++i) {
 		if (!counter_info [i])
 			goto failure;
 		size += strlen (counter_info [i]) + 1;
 	}
+
 	size += 7;
 	size &= ~7;
+	int counters_data_size;
 	counters_data_size = num_counters * 8; /* optimize for size later */
 	if (size > 65535)
 		goto failure;
 	perfctr_lock ();
+	unlock = TRUE;
+	SharedCategory *cat;
 	cat = (SharedCategory*) shared_data_reserve_room (size, FTYPE_CATEGORY);
-	if (!cat) {
-		perfctr_unlock ();
+	if (!cat)
 		goto failure;
-	}
 	cat->num_counters = num_counters;
 	cat->counters_data_size = counters_data_size;
-	/* now copy the vaiable data */
+	/* now copy the variable data */
+	char *p;
 	p = cat->name;
-	strcpy (p, name);
-	p += strlen (name) + 1;
-	strcpy (p, chelp);
-	p += strlen (chelp) + 1;
+	p = append (p, name);
+	p = append (p, chelp);
+
 	for (i = 0; i < num_counters; ++i) {
-		CounterCreationData *data = mono_array_get (items, CounterCreationData*, i);
+		MONO_HANDLE_LOCAL_VARIABLE_INITIALIZED_NULL (CounterCreationData, data);
```

Comment:
Here we allocate a coop handle once per loop iteration.  `num_counters` is controlled by user code (it's the length of the `items` array.  One of the requirements for coop handles is that the number of live handles must not be unbounded.

In cases where we need to loop over user arguments, the solution we've come up with is to pull the loop body in to a helper function that uses `HANDLE_FUNCTION_ENTER/RETURN` to control the lifetime of the coop handles needed by the body.

I know that seems like an ugly and unsatisfying pattern, but it seemed less invasive and error-prone that anything else we tried.